src/NNCache.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+    This file is part of Leela Zero.
+    Copyright (C) 2017 Michael O
+
+    Leela Zero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Leela Zero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "config.h"
+#include <functional>
+
+#include "NNCache.h"
+#include "Utils.h"
+
+NNCache::NNCache(int size) : m_size(size) {}
+
+NNCache& NNCache::get_NNCache(void) {
+    static NNCache cache;
+    return cache;
+}
+
+bool NNCache::lookup(Key hash, Network::Netresult & result) {
+#ifndef NDEBUG
+    if (m_lookups % 10000 == 0) {
+        dump_stats();
+    }
+#endif
+    std::lock_guard<std::mutex> lock(m_mutex);
+    ++m_lookups;
+
+    auto iter = m_cache.find(hash);
+    if (iter == m_cache.end()) {
+        return false;  // Not found.
+    }
+
+    const auto& entry = iter->second;
+
+    // Found it.
+    ++m_hits;
+    result = entry->result;
+    return true;
+}
+
+void NNCache::insert(Key hash,
+                     const Network::Netresult& result) {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    if (m_cache.find(hash) != m_cache.end()) {
+        return;  // Already in the cache.
+    }
+
+    m_cache.emplace(hash, std::make_unique<Entry>(result));
```

Comment:
This does two lookups for new entries (once in find() and once here). You could always do just one, e.g.:
```c++
auto result = m_cache.emplace(hash, nullptr);
if (!result.second) {
  return;  // Already in the cache
}

result.first->second = std::make_unique<Entry>(result);
```

Reply:
We need to protect against races when there are multiple threads. The lock is not held during the time Network is running because that takes a very long time.

Reply:
Granted, but you already have a lock at line 56. I wasn't suggesting to remove it, just to replace find+emplace with a single emplace.

Reply:
Oh I get it now. I'm pretty new to C++ so I'm looking at the reference for std::map::emplace. Also I think your code `auto result` is colliding with the insert parameter result? That really confused me when I first read it.

I think this is a good idea but maybe I'll wait until later to do it. BTW if you're ambitious how about doing a pull request for this to the LZGo project, they have the exact same code, I just copied this verbatim from there and made minor change to adapt it.

Reply:
Whoops, "result" was indeed an unfortunate name for the variable. Sorry about the confusion!

No problem, it's just a tiny optimization. I'm sure it's dwarfed by actual NN evals.